File: 08-streaming/8.4-structured/src/main/scala/x/StructuredStreaming.scala
package x


import org.apache.spark.sql.functions._
import org.apache.spark.sql.SparkSession
//import spark.implicits._

/*
$  spark-submit  --master local[2]   --driver-class-path logging/  --class x.StructuredStreaming  target/scala-2.11/structured-streaming_2.11-1.0.jar
 */


object StructuredStreaming {
  def main(args: Array[String]) {

    val spark = SparkSession.builder.appName("Structured Streaming").
                getOrCreate()

    // TODO-1  : set the port to 10000
    val clickstream = spark.readStream.format("socket").
                      option("host", "localhost").
                      option("port", ???)
                      .load()

    // TODO-2 : printSchema
    clickstream.???

    val query = clickstream.writeStream.
                outputMode("append")
                .format("console")
                .start()

    query.awaitTermination()

  }
}

File: 08-streaming/8.4-structured/python/structured-streaming.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import explode
from pyspark.sql.functions import split

# Initialize Spark
spark = SparkSession \
    .builder \
    .appName("Structed Streaming") \
    .getOrCreate()

# Set loglevel to Error
spark.sparkContext.setLogLevel("ERROR")


## TODO-1 : read from socket 10000
lines = spark \
    .readStream \
    .format("socket") \
    .option("host", "localhost") \
    .option("port", 10000) \
    .load()

# query1 = lines \
#     .writeStream \
#     .outputMode("append") \
#     .format("console") \
#     .start()

# query1.awaitTermination()



## TODO-2  :filter lines that has 'x'
x = lines.filter(lines["value"].contains("x"))

## TODO-3 : To run query2 , comment out query1
query2 = x \
        .writeStream \
        .outputMode("append") \
        .format("console") \
        .start()
query2.awaitTermination()


# wait forever until user terminate manually
spark.streams.awaitAnyTermination() 
spark.stop()
